lib/Dialect/Triton/Transforms/Combine.cpp

```diff
@@ -207,12 +244,16 @@ class CombineOpsPass : public TritonCombineOpsBase<CombineOpsPass> {
     // patterns.add<CombineAddPtrPattern>(context);
     patterns.add<CombineBroadcastConstantPattern>(context);
     patterns.add<CombineBroadcastMulReducePattern>(context);
+    patterns.add<CombineNopTransPattern>(context);
```

Comment:
I think those 2 should be canonicalization instead

Reply:
How is this different from the other combine rules here?

(I thought of canonicalization as things like `add imm + x` -> `add x + imm`.  This seems not like that, since it's either removing the instr entirely, or combining two instrs?)

I wonder if it should be a fold instead?  That seems better because it should be faster than a rewrite pattern?

Reply:
you’re right, they should both be folder

Reply:
`transpose(x, order=[0, 1, ...]) -> x` is indeed a folding case.
`transpose(transpose(x)) -> transpose(x)` seems like a more general canonicalization but not folding since "fold has the restriction that no new operations may be created"

Reply:
But fold is allowed to modify the instruction, so can't I change the outer transpose?

Reply:
Got it. Maybe you can update in-place.
We define other patterns (e.g., view(view) -> view) as a canonicalization instead of folding in triton.

Reply:
correct fold allow to either replace the op by a Value or mutate the op. In this case the first one would just return the source and the second case would mutate the transpose. I think they can both fit into MLIR folding.

> (I thought of canonicalization as things like add imm + x -> add x + imm. This seems not like that, since it's either removing the instr entirely, or combining two instrs?)

BTW in MLIR canonicalization has a bit of a different meaning which includes all the basic patterns that are not folding but basic rewrites

Reply:
Here is an example:
https://github.com/llvm/llvm-project/blob/ff05c3087b4b7e0121125d7f6860b3d9383ca91d/mlir/lib/Dialect/Arith/IR/ArithOps.cpp#L1270

Reply:
Maybe it's worth revising our existing patterns like `view(view) -> view` with folding then as it's faster?

Reply:
+1 that would be a nice cleanup (not in this PR of course), folding also runs in more cases which is convenient.

Reply:
If this is the only outstanding comment, how would you feel if I checked this in as-is and then cleaned up all of the patterns in a follow-up?  Or is there harm in having this as a non-canonicalization non-fold?

Reply:
it's a nice improvement but definitely not a deal breaker so I'll leave it up to you

Reply:
Thanks -- I definitely do want to change it all to use `fold`, but this patch is big enough as-is.  I'll work on fold now.

Reply:
> Maybe it's worth revising our existing patterns like view(view) -> view with folding then as it's faster?

@Jokeren Where do you see this?  I see foldViewLikeOp in Ops.cpp, but that's a fold.

Now that I look I actually don't see anything else we can change to `fold`, other than what I introduced in this patch.  (In which case, sorry, I should have just changed it here.)  Do you all see anything else?

Reply:
https://github.com/openai/triton/pull/3029

Reply:
How about this one:
https://github.com/openai/triton/blob/26e2e2d2d5a218a3cd7cf60f8a1f66feb6c200b0/lib/Dialect/Triton/IR/Ops.cpp#L811

Reply:
Aha, I missed that one because the code said replaceWithNewOp.

I went and wrote the patch, but looking at it, I actually don't like it, because we have to duplicate the condition on reshape, `  if (!op.getAllowReorder() || op.getEfficientLayout().has_value())`.

https://gist.github.com/jlebar/d183a6c9de62fe7d9c2f85c3c58d6f98

I think I'll leave it as-is unless you feel strongly.